ott/geometry/graph.py

```diff
@@ -0,0 +1,328 @@
+from typing import Any, Dict, Optional, Sequence, Tuple, Union
+
+import jax
+import jax.experimental.sparse as jesp
+import jax.numpy as jnp
+from typing_extensions import Literal
+
+from ott.core import decomposition, fixed_point_loop
+from ott.geometry import geometry
+
+__all__ = ["Graph"]
+
+Sparse_t = Union[jesp.CSR, jesp.CSC, jesp.COO, jesp.BCOO]
+
+
+@jax.tree_util.register_pytree_node_class
+class Graph(geometry.Geometry):
+  r"""Graph distance approximation using heat kernel :cite:`heitz:21,crane:13`.
+
+  Approximates the heat kernel for large ``n_steps``, which for small ``t``
+  approximates the geodesic exponential kernel :math:`e^{\frac{-d(x, y)^2}{t}}`.
+
+  For sparse graphs, :mod:`sksparse.cholmod` is required to compute the Cholesky
+  decomposition.
+  Differentiating w.r.t. the edge weights is currently possible only when the
+  graph is represented as a dense adjacency matrix.
+
+  Args:
+    graph: Graph represented as an adjacency matrix of shape ``[n, n]``.
+      If `None`, the symmetric graph Laplacian has to be specified.
+    laplacian: Symmetric graph Laplacian. The check for symmetry is **NOT**
+      performed. If `None`, the graph has to be specified instead.
+    t: Constant used when approximating the geodesic exponential kernel.
+      If `None`, use :math:`\frac{1}{|E|} \sum_{(u, v) \in E} weight(u, v)`
+      :cite:`crane:13`. In this case, the ``graph`` must be specified
+      and the edge weights are all assumed to be positive.
+    n_steps: Maximum number of steps used to approximate the heat kernel.
+    numerical_scheme: Numerical scheme used to solve the heat diffusion.
+    directed: Whether the ``graph`` is directed. If not, it will be made
+      undirected as :math:`G + G^T`. This parameter is ignored when  directly
+      passing the Laplacian, which is assumed to be symmetric.
+    tol: Relative tolerance with respect to the Hilbert metric, see
```

Comment:
@marcocuturi am not sure about this option, would consider removing it since it doesn't seem to help that much (at least in tests), but am happy to discuss.

Reply:
does it work better now with re-entering?

Reply:
Don't know yet, still need to test it out.